### PR TITLE
Add 400 response when `boundary` is missing 

### LINKED ISSUE #1588: Send 400 on missing `boundary`
I did some research about this PR, and to understand if the issue was reasonable.

Here's what I've found out:
- Flask raises `ValueError` when the `boundary` is not found: https://github.com/pallets/werkzeug/blob/dae7e0d06651e54a74f04e1cf24d806c4e2e9be9/src/werkzeug/formparser.py#L270-L291
- Django raises a custom exception: https://github.com/django/django/blob/abfdb4d7f384fb06ed9b7ca37b548542df7b5dda/django/http/multipartparser.py#L79-L83
- Django also verifies if the `boundary` is valid with https://github.com/python/cpython/blob/27ee43183437c473725eba00def0ea7647688926/Lib/cgi.py#L991-L997
- After that raise, Django catches that exception and raises a 400: https://github.com/django/django/blob/abfdb4d7f384fb06ed9b7ca37b548542df7b5dda/django/core/handlers/exception.py#L84-L94

On Starlette, we raise `KeyError` with "boundary key missing". Although this PR follows the issue that was raised, and it's actually an improvement, I think we should follow Django's lead, and also create a 400 response on this case.

_Originally posted by @Kludex in https://github.com/encode/starlette/issues/1544#issuecomment-1080197920_

### PR DESCRIPTION
**Pull request by Kludex**
- Closes #1588
- Related to #1303 (I've not included it on this PR for now...)
- Replaces #1593 (monologue about possible solutions from there is important)

Behavior of this PR:
- Custom exception in case `"app" not in scope`.
- `HTTPException(status_code=400)` in case `"app" in scope`.

Changes:
- Move `HTTPException` to `http_exception.py` to avoid circular import.
- Add custom exception `MultiPartException` on missing `boundary`.
- Handle `MultiPartException` on `form()` with the behavior mentioned above.

Alternatives/thoughts related to this PR:
- Check #1593 - my thoughts are listed there in a timely manner...
- The handling of `MultiPartException` can also be on an upper level:
https://github.com/encode/starlette/blob/830f3486537916bae6b46948ff922adc14a22b7c/starlette/routing.py#L250-L261
This alternative would avoid creating the `http_exception.py`... 🤔

### CODE REVIEW

File: starlette/http_exception.py
Comment (adriangb) [inline starlette/http_exception.py]: Why do we need to move this?

File: starlette/formparsers.py
Comment (adriangb) [inline starlette/formparsers.py]: Why do we need to define `__init__` here? Are we using `*args`?

File: starlette/formparsers.py
Comment (Kludex) [inline starlette/formparsers.py]:
```suggestion
   ...
```

File: starlette/http_exception.py
Comment (Kludex) [reply starlette/http_exception.py]: Circular import. `exceptions -> requests`; `requests -> formparsers`; `formparsers -> exceptions`

Comment (adriangb) [reply starlette/http_exception.py]: Gotcha. I was always confused as to why there was a middleware in the exceptions module. Could we move the middleware instead?

Comment (Kludex) [reply starlette/http_exception.py]: hmm... I didn't think about it. But it makes sense. 👍

File: starlette/formparsers.py
Comment (Kludex) [inline starlette/formparsers.py]:
```suggestion
    ...
```

File: starlette/http_exception.py
Comment (adriangb) [reply starlette/http_exception.py]: might be nice to "deprecate" it from this module as well even if we re-export it in the short term

Comment (Kludex) [reply starlette/http_exception.py]: Ok. That's probably the reason I didn't thought about it. With the relocation of `HTTPException`, the import remains here, so there's no need to deprecate it. I'd have to introduce the `__dir__` and `__getattr__` on this file as well.

File: starlette/formparsers.py
Comment (Kludex) [reply starlette/formparsers.py]: No, only the message. I've removed when you mentioned this, but I need to put it back. Conceptually, `MultiPartException` is not an `HTTPException`, so it needs to inherit from `Exception`. The message is necessary here because there will be another part of the code that will raise this error in a future PR (the "name" mentioned on the description).

File: starlette/http_exception.py
Comment (adriangb) [reply starlette/http_exception.py]: I think that's worth it right? I really do feel like the middleware should be in `starlette.middleware.exceptions`

File: starlette/formparsers.py
Comment (adriangb) [reply starlette/formparsers.py]: I still don't understand why it needs to be here, standard behavior for exceptions is that the first argument is a string which corresponds to the exception message. Is this just for when `"app"` is not in scope? If so users can just parse the exceptions, I don't think it's a use case we should be putting effort into supporting.

File: starlette/requests.py
Comment (Kludex) [inline starlette/requests.py:259]: I don't want to use `exc.args[0]` here. Do you have a better solution? @adriangb

File: starlette/formparsers.py
Comment (Kludex) [reply starlette/formparsers.py]: Let's continue on the other thread.

File: starlette/requests.py
Comment (Kludex) [reply starlette/requests.py:259]: I can have two different exceptions as well, and don't use `exc.message`...

Comment (adriangb) [reply starlette/requests.py:259]: Ah I see where this was going now! Sorry for not catching on. Seeing this I think that `message` is okay, but 2 exceptions are also a good idea, I think there's precedent for both. I'll leave it up to you.

File: starlette/http_exception.py
Comment (Kludex) [reply starlette/http_exception.py]: Not sure 😅

File: starlette/exceptions.py
Comment (Kludex) [inline starlette/exceptions.py:29]: I need to add a test for this.

File: starlette/http_exception.py
Comment (adriangb) [reply starlette/http_exception.py]: Fair enough, it's your call now that you've tried it both ways!

File: tests/test_formparsers.py
Comment (adriangb) [inline tests/test_formparsers.py:401]: The use of `does_not_raise()` here is super nifty!

File: starlette/exceptions.py
Comment (adriangb) [inline starlette/exceptions.py:29]: What's the plan to remove all of this deprecation code? I get why we need it, but it'd be nice to know when we're going to remove it so that:

1. We can put it in the code for our knowledge down the road.
2. We can put it in the error message.

I suppose this depends on #1623 , but I think we should make an executive decision (maybe @tomchristie ?) on when / if the deprecation shim is ever going to be removed.

Personally, I feel like 3 minor releases or 1 year, whichever comes first, sounds reasonable?

**Review by adriangb — APPROVED** [review]

**Review by adriangb — APPROVED** [review]
This is an interesting change.

Overall the change (returning a proper exception to the client instead of a 500) makes absolute sense.

Unfortunately you ended up needing to do a bunch of surgery to the codebase to make it happen because raising an `HTTPexception` from within a `Request` requires that `Request` reference `HTTPException`, but `ExceptionMiddleware` was in the same module as `HTTPException`, which caused circular references. This makes the diff super ugly and huge relative to the small straightforward feature.

But I think that's okay: we are paying the price to fix an issue that already existed in our codebase that might have caused us issues down the road some other time. And personally I think moving the middleware into `middlewares/` really cleans things up nicely.

I'm approving this, I do however want 1 more approver and some consensus on the deprecation policy because this is a breaking change for our downstream dependencies.

File: starlette/exceptions.py
Comment (Kludex) [reply starlette/exceptions.py:29]: I was more into 6 months and 3 minors... But any number is good for me... I'd just like to have a number 👍 But let's mention those stuff on that discussion.

For here, there's no plan yet. The plan should be defined on that discussion. If this is merged, it's just common sense i.e. in some arbitrary time we just remove it...

**Comment by Kludex** [comment]
> [...] I do however want 1 more approver [...]

Yep 👍 

Thanks for the review! 🙏

File: starlette/exceptions.py
Comment (lovelydinosaur) [reply starlette/exceptions.py:29]: Didn't realise you could do this at a module level. 😎

File: starlette/middleware/exceptions.py
Comment (lovelydinosaur) [inline starlette/middleware/exceptions.py:11]: 👍 Yup - much better to have this here.

Have confirmed to myself that the class here and the class when it was in `exceptions.py` are identical.

File: starlette/requests.py
Comment (lovelydinosaur) [inline starlette/requests.py:258]: Do we need to have the switch here? Can we just *always* coerce to an `HTTPException` here instead? Are we already using this pattern, and if so, where?

File: tests/test_formparsers.py
Comment (lovelydinosaur) [inline tests/test_formparsers.py:408]: I'd *personally* probably nudge towards the simpler `pytest.raises(KeyError, match="boundary")` case here, just because it's more obviously readable to me.

The parameterised `expectation` is neat, but also more complex to understand.

Just my personal perspective tho, happy to go either way.

**Review by lovelydinosaur — APPROVED** [review]
Some questions which could be addressed, but they're not blockers so will leave this in your hands at this point @Kludex.

Really nice switch around with moving the middleware. 👍

File: starlette/requests.py
Comment (adriangb) [reply starlette/requests.py:258]: https://github.com/encode/starlette/blob/master/starlette/routing.py#L253-L258

Comment (Kludex) [reply starlette/requests.py:258]: Yep, we use the pattern in the code, as @adriangb pointed out.

Always coerce to an `HTTPException` is an option. I opted to not do it because we only use `HTTPException`s when we are in `Starlette` (as application) e.g. `"app" in self.scope`.

File: tests/test_formparsers.py
Comment (Kludex) [reply tests/test_formparsers.py:408]: The idea was to avoid the creation of two very similar tests. Do you think is clear if I create two tests instead?

File: starlette/requests.py
Comment (adriangb) [reply starlette/requests.py:258]: Realistically, I think we could remove that pattern everywhere. IMO it's not worth supporting (nice) usage of Starlette's Request object outside of a Starlette app. If someone wants to do that, they can handle the HTTPException.

**Comment by adriangb** [comment]
I think your way or making 1 test is the cleanest I've seen, but if it's
only 2 tests and not 10 I also think making it more explicit would be nice.

File: tests/test_formparsers.py
Comment (adriangb) [reply tests/test_formparsers.py:408]: I think your way or combining an exception and non-exception test into 1 is the cleanest I've seen for that pattern, but if it's
only 2 tests and not 10 I also think making it more explicit would be nice

File: tests/test_formparsers.py
Comment (Kludex) [reply tests/test_formparsers.py:401]: Thanks! 🙇

File: starlette/requests.py
Comment (Kludex) [reply starlette/requests.py:258]: >  If someone wants to do that, they can handle the HTTPException.

You have a point, but I think that's another discussion.

I really don't think it's a burden to maintain that, but I wonder if someone really uses that code outside `Starlette` (app)... 🤔

**Comment by Kludex** [comment]
I'm going to merge this.

Some things left unanswered:
- Deprecation policy, how do we proceed?
- Should we check if someone is using parts of Starlette without the Starlette application itself? i.e. `if "app" in scope:`.

**Pull request merged**
